### PR TITLE
feat: Adding constraint template and constraints to verify deprecated API

### PR DESCRIPTION
**Pull request by JaydipGabani**
Signed-off-by: Jaydip Gabani <gabanijaydip@gmail.com>

**What this PR does / why we need it**:
Adds a constraint template and constraint to audit and verify deprecated API versions.

**Which issue(s) does this PR fix** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by JaydipGabani** [comment]
@sozercan I have used [this guide](https://kubernetes.io/docs/reference/using-api/deprecation-guide/) as source of truth for constraints.

**Comment by maxsmythe** [comment]
Is this intended to be run in audit or purely as a webhook?

**Comment by JaydipGabani** [comment]
Both!

**Comment by maxsmythe** [comment]
One issue with audit:

If I kubectl apply a Pod named `my-favorite-server` as v1, Kubernetes takes that request and serializes it into etcd using a storage version. If I run `kubectl get` I can get either a v1 version or a v1beta1 version.

You can try this yourself in a cluster with gatekeeper installed by running:

`kubectl get validatingwebhookconfigurations.v1.admissionregistration.k8s.io -oyaml gatekeeper-validating-webhook-configuration`

and

`kubectl get validatingwebhookconfigurations.v1beta1.admissionregistration.k8s.io -oyaml gatekeeper-validating-webhook-configuration`

Note that the API version changes (I'm not sure how much conversion there is between versions for VWH configs).

Audit essentially does the same thing: it pulls every object from the cluster in every available API version, so the audit version of this constraint wouldn't necessarily test usage of a deprecated API, but only its existence.

Sadly, usage is undetectable by audit, as the original resource used to create an object is not available.

https://github.com/open-policy-agent/gatekeeper/issues/2117 has more context

**Comment by ctrought** [comment]
If the original API version is still available by the team it reaches GK webhook it might be an interesting use case for further enhancement of Mutation APIs. Imagine being able to store the original API version as an annotation when the object is first applied so that subsequent audit run constraints could access it after it were applied (assuming it is a warning and not denied). Of course it would only work for objects already existing in the cluster that have an update applied to it or deleted/recreated.

**Comment by JaydipGabani** [comment]
@maxsmythe sounds like we want this constraint for admission only then 🤔 . If so, this we be dependant on - [Gatekeeper constraint per process enforcement actions - Google Docs](https://docs.google.com/document/d/1QIABjZN9B8oBZF5mm1hz5IhpTGulm1jTNn20nYIF-Hc/edit#heading=h.9oaxevykg8w6)

File: artifacthub/library/general/verifydepricatedapi/1.0.0/artifacthub-pkg.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+version: 1.0.0
+name: verifydepricatedapi
```
Comment (ctrought) [inline artifacthub/library/general/verifydepricatedapi/1.0.0/artifacthub-pkg.yml:2]: nit: deprecated*

**Comment by maxsmythe** [comment]
> On another thought, this constraint template won't affect the audit anyways, for the resources with API versions that are deprecated cannot exist in the cluster.

Then they would be rejected by the API server as well, so there would be no need for them. Or is this only for `gator test`?

**Comment by maxsmythe** [comment]
@ctrought 

>  Of course it would only work for objects already existing in the cluster that have an update applied to it

Yeah, the difficulty with auditing something like this is there may be multiple actors working on the same object, using different versions to read/write from it. K8s audit logs can show per-actor changes. Mutation might be able to show timestamps/usernames for writes using various API versions.

For audit and both validation and mutation, reads would be unobservable.

**Comment by JaydipGabani** [comment]
> > On another thought, this constraint template won't affect the audit anyways, for the resources with API versions that are deprecated cannot exist in the cluster.
> 
> Then they would be rejected by the API server as well, so there would be no need for them. Or is this only for `gator test`?

I changed the enforcement action to `warn` as this constraint template should provide a warning and not deny the resource. 

API server wouldn't reject or warn users against future depreciations, right? It is meant to be used for that. Imagine a situation where I am using the 1.25 version and creating resources that are deprecating in 1.27, API server won't warn me against deprecating resources in 1.27 by default right? This constraint template can be used to warn users about deprecating resources when API server fails to.

File: artifacthub/library/general/verifydeprecatedapi/1.0.0/template.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        get_message(kind, apiVersion, k8sVersion, targetAPI) = msg {
+          not match(targetAPI)
+          msg := sprintf("API %v for %v is deprecated in kubeversion %v, please use %v instead", [kind, apiVersion, k8sVersion, targetAPI])
+        }
+
+        get_message(kind, apiVersion, k8sVersion, targetAPI) = msg {
+          match(targetAPI)
+          msg := sprintf("API %v for %v is deprecated in kubeversion %v, please see kubernetes API deprication guide", [kind, apiVersion, k8sVersion])
```
Comment (sozercan) [inline artifacthub/library/general/verifydeprecatedapi/1.0.0/template.yaml:59]:
```suggestion
          msg := sprintf("API %v for %v is deprecated in Kubernetes version %v, please see Kubernetes API deprecation guide", [kind, apiVersion, k8sVersion])
```

File: library/general/verifydeprecatedapi/template.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        get_message(kind, apiVersion, k8sVersion, targetAPI) = msg {
+          not match(targetAPI)
+          msg := sprintf("API %v for %v is deprecated in kubeversion %v, please use %v instead", [kind, apiVersion, k8sVersion, targetAPI])
+        }
+
+        get_message(kind, apiVersion, k8sVersion, targetAPI) = msg {
+          match(targetAPI)
+          msg := sprintf("API %v for %v is deprecated in kubeversion %v, please see kubernetes API deprication guide", [kind, apiVersion, k8sVersion])
```
Comment (sozercan) [inline library/general/verifydeprecatedapi/template.yaml:59]:
```suggestion
          msg := sprintf("API %v for %v is deprecated in Kubernetes version %v, please see Kubernetes API deprecation guide", [kind, apiVersion, k8sVersion])
```

File: src/general/verifydeprecatedapi/src.rego [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+get_message(kind, apiVersion, k8sVersion, targetAPI) = msg {
+  not match(targetAPI)
+  msg := sprintf("API %v for %v is deprecated in kubeversion %v, please use %v instead", [kind, apiVersion, k8sVersion, targetAPI])
+}
+
+get_message(kind, apiVersion, k8sVersion, targetAPI) = msg {
+  match(targetAPI)
+  msg := sprintf("API %v for %v is deprecated in kubeversion %v, please see kubernetes API deprication guide", [kind, apiVersion, k8sVersion])
```
Comment (sozercan) [inline src/general/verifydeprecatedapi/src.rego:18]:
```suggestion
          msg := sprintf("API %v for %v is deprecated in Kubernetes version %v, please see Kubernetes API deprecation guide", [kind, apiVersion, k8sVersion])
```

File: website/docs/verifydeprecatedapi.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+        }
+
+        get_message(kind, apiVersion, k8sVersion, targetAPI) = msg {
+          not match(targetAPI)
+          msg := sprintf("API %v for %v is deprecated in kubeversion %v, please use %v instead", [kind, apiVersion, k8sVersion, targetAPI])
+        }
+
+        get_message(kind, apiVersion, k8sVersion, targetAPI) = msg {
+          match(targetAPI)
+          msg := sprintf("API %v for %v is deprecated in kubeversion %v, please see kubernetes API deprication guide", [kind, apiVersion, k8sVersion])
```
Comment (sozercan) [inline website/docs/verifydeprecatedapi.md:71]:
```suggestion
          msg := sprintf("API %v for %v is deprecated in Kubernetes version %v, please see Kubernetes API deprecation guide", [kind, apiVersion, k8sVersion])
```

File: library/general/verifydeprecatedapi/suite.yaml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+kind: Suite
+apiVersion: test.gatekeeper.sh/v1alpha1
+metadata:
+  name: verifydeprecatedapi
+tests:
+- name: block-endpoint-default-role
```
Comment (sozercan) [inline library/general/verifydeprecatedapi/suite.yaml:6]: looks like this needs to be changed

File: library/general/verifydeprecatedapi/suite.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  constraint: samples/verifydeprecatedapi/constraint.yaml
+  cases:
+  - name: example-allowed
+    object: samples/verifydeprecatedapi/example_allowed.yaml
+    assertions:
+    - violations: no
+  - name: example-disallowed
+    object: samples/verifydeprecatedapi/example_disallowed.yaml
+    assertions:
+    - violations: yes
```
Comment (sozercan) [inline library/general/verifydeprecatedapi/suite.yaml:17]: can we add tests for other versions too

File: library/general/verifydeprecatedapi/template.yaml [hunk 8]
```diff
@@ -0,0 +1,64 @@
+apiVersion: templates.gatekeeper.sh/v1
+kind: ConstraintTemplate
+metadata:
+  name: verifydeprecatedapi
+  annotations:
+    metadata.gatekeeper.sh/title: "Verify deprecated APIs"
+    metadata.gatekeeper.sh/version: 1.0.0
+    description: >-
+      Verifies deprecated kuberentes APIs to insure all the API versions are upto date.
```
Comment (sozercan) [inline library/general/verifydeprecatedapi/template.yaml:9]:
```suggestion
      Verifies deprecated Kubernetes APIs to ensure all the API versions are up to date.
```

File: library/general/verifydeprecatedapi/template.yaml [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+                    type: string
+                    description: deprecated api
+                  kinds: 
+                    type: array
+                    items:
+                      type: string
+                    description: impacted list of kinds  
+                  targetAPI:
+                    type: string
+                    description: target api
```
Comment (sozercan) [inline library/general/verifydeprecatedapi/template.yaml:35]:
```suggestion
                    description: target api
```

File: library/general/verifydeprecatedapi/template.yaml [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+                type: object
+                properties:
+                  deprecatedAPI:
+                    type: string
+                    description: deprecated api
+                  kinds: 
+                    type: array
+                    items:
+                      type: string
+                    description: impacted list of kinds
```
Comment (sozercan) [inline library/general/verifydeprecatedapi/template.yaml:32]:
```suggestion
                    description: impacted list of kinds
```

File: library/general/verifydeprecatedapi/template.yaml [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    spec:
+      names:
+        kind: VerifyDeprecatedAPI
+      validation:
+        openAPIV3Schema:
+          type: object
+          properties:
+            kvs:
+              type: array
+              description: Deprecated api versions and corrosponding kinds
```
Comment (sozercan) [inline library/general/verifydeprecatedapi/template.yaml:21]:
```suggestion
              description: Deprecated API versions and corresponding kinds
```

File: library/general/verifydeprecatedapi/suite.yaml (hunk 7 above)
Comment (JaydipGabani) [reply library/general/verifydeprecatedapi/suite.yaml:17]: sure!

**Comment by sozercan** [comment]
I think scenarios for this would be gator and webhook (mostly deny for future deprecated version versions)

**Comment by JaydipGabani** [comment]
@sozercan sorry for the delay! Added tests for other Kubernetes versions as well.

**Review by sozercan — APPROVED** [review]
LGTM

File: artifacthub/library/general/verifydeprecatedapi/1.0.0/samples/verify-1.16/example_allowed.yaml [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - name: nginx
+        image: nginx:1.14.2
+        ports:
+        - containerPort: 80
```
Comment (maxsmythe) [inline artifacthub/library/general/verifydeprecatedapi/1.0.0/samples/verify-1.16/example_allowed.yaml:21]: nit: newlines (for multiple files)

File: library/general/verifydeprecatedapi/template.yaml [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+          properties:
+            kvs:
+              type: array
+              description: Deprecated api versions and corresponding kinds
+              items:
+                type: object
+                properties:
+                  deprecatedAPI:
+                    type: string
+                    description: deprecated api
```
Comment (maxsmythe) [inline library/general/verifydeprecatedapi/template.yaml:27]: an example value here would help clarify the expected string format

**Review by maxsmythe — APPROVED** [review]
LGTM with some nits.

The audit story for this constraint will be fraught (until we can exclude constraints from audit)

File: website/docs/verifydeprecatedapi.md [hunk 14]
```diff
@@ -0,0 +1,759 @@
+---
+id: verifydeprecatedapi
+title: Verify deprecated APIs
+---
+
+# Verify deprecated APIs
+
+## Description
+Verifies deprecated Kubernetes APIs to ensure all the API versions are up to date.
+
```
Comment (sozercan) [inline website/docs/verifydeprecatedapi.md:10]: can we add note saying this won't be applicable for audit and state the reason

Comment (JaydipGabani) [reply website/docs/verifydeprecatedapi.md:10]: done!

**Pull request merged**